build_system/seqan3-config.cmake

```diff
@@ -112,51 +112,65 @@ macro (seqan3_config_error text)
+    set (SEQAN3_HAS_CPM TRUE)
+else ()
+    set (SEQAN3_HAS_CPM FALSE)
+endif ()
+
+if (SEQAN3_HAS_CPM)
+    set (CPM_INDENT "  CMake Package Manager CPM: ")
+    include ("${CMAKE_CURRENT_LIST_DIR}/CPM.cmake")
+    CPMUsePackageLock ("${CMAKE_CURRENT_LIST_DIR}/package-lock.cmake")
+endif ()
+
 # ----------------------------------------------------------------------------
 # Find SeqAn3 include path
 # ----------------------------------------------------------------------------
 
 # Note that seqan3-config.cmake can be standalone and thus SEQAN3_CLONE_DIR might be empty.
-# * `SEQAN3_CLONE_DIR` was already found in seqan3-config-version.cmake
 # * `SEQAN3_INCLUDE_DIR` was already found in seqan3-config-version.cmake
-find_path (SEQAN3_SUBMODULES_DIR
-           NAMES submodules/sdsl-lite
-           HINTS "${SEQAN3_CLONE_DIR}" "${SEQAN3_INCLUDE_DIR}/seqan3")
-
 if (SEQAN3_INCLUDE_DIR)
     seqan3_config_print ("SeqAn3 include dir found:   ${SEQAN3_INCLUDE_DIR}")
 else ()
     seqan3_config_error ("SeqAn3 include directory could not be found (SEQAN3_INCLUDE_DIR: '${SEQAN3_INCLUDE_DIR}')")
 endif ()
 
 # ----------------------------------------------------------------------------
-# Detect if we are a clone of repository and if yes auto-add submodules
+# Require SDSL
 # ----------------------------------------------------------------------------
 
-if (SEQAN3_CLONE_DIR)
-    seqan3_config_print ("Detected as running from a repository checkout…")
-endif ()
+find_path (SEQAN3_SDSL_INCLUDE_DIR
+           NAMES sdsl/version.hpp
+           HINTS "${SEQAN3_INCLUDE_DIR}/seqan3/vendor")
 
-if (SEQAN3_SUBMODULES_DIR)
-    file (GLOB submodules ${SEQAN3_SUBMODULES_DIR}/submodules/*/include)
-    foreach (submodule ${submodules})
-        if (IS_DIRECTORY ${submodule})
-            seqan3_config_print ("  …adding submodule include:  ${submodule}")
-            set (SEQAN3_DEPENDENCY_INCLUDE_DIRS ${submodule} ${SEQAN3_DEPENDENCY_INCLUDE_DIRS})
-        endif ()
-    endforeach ()
+# 1) Check the vendor directory of SeqAn3. This directory exists for source packages and installed packages.
+if (SEQAN3_SDSL_INCLUDE_DIR)
+    seqan3_config_print ("Required dependency:        SDSL found.")
+    set (SEQAN3_DEPENDENCY_INCLUDE_DIRS ${SEQAN3_SDSL_INCLUDE_DIR} ${SEQAN3_DEPENDENCY_INCLUDE_DIRS})
+    # 2) Get package via CPM.
+elseif (SEQAN3_HAS_CPM)
+    CPMGetPackage (sdsl-lite)
+
+    find_path (SEQAN3_SDSL_INCLUDE_DIR
+               NAMES sdsl/version.hpp
+               HINTS "${sdsl-lite_SOURCE_DIR}/include")
+
+    if (SEQAN3_SDSL_INCLUDE_DIR)
+        seqan3_config_print ("Required dependency:        SDSL found.")
+        set (SEQAN3_DEPENDENCY_INCLUDE_DIRS ${SEQAN3_SDSL_INCLUDE_DIR} ${SEQAN3_DEPENDENCY_INCLUDE_DIRS})
+    else ()
+        seqan3_config_error ("The SDSL library is required, but wasn't found.")
+    endif ()
+else ()
+    seqan3_config_error ("The SDSL library is required, but wasn't found.")
```

Comment:
This whole block seems twisted.

Reply:
Yes, because it needs to work for clones, source packages, and installed packages.
Ideally, there would be different scripts for those three, but it would require some rewriting of all the CMake stuff.